internal/components/registry.go

```diff
@@ -52,7 +47,12 @@ func (r *Registry) addModels(models []Coremodel) error {
 
 	// Update model index and return an error if trying to register a duplicate.
 	for _, m := range models {
-		k := makeRegistryKey(m.Schema())
+		k := m.Lineage().Name()
+
+		// Ensure assignability first. TODO will this blow up for dashboards?
+		if err := thema.AssignableTo(m.Current(), m.GoType()); err != nil {
+			return fmt.Errorf("%s schema version %v not assignable to provided Go type: %w", k, m.Current().Version(), err)
+		}
```

Comment:
I don't think this belongs to the Registry. Maybe we could do this in a loader instead?

Reply:
i went back and forth on this a bit, but it ultimately came down to exposing the standard Thema `LineageFactory` func. As long as that's a goal, then there can't really be a single loader which both loads a lineage _and_ checks Go struct assignability, as your `loadThemaLineage()` does/did, because the logical layer of `LineageFactory` does not assume the existence of a corresponding Go struct.

If/when we genericize the clustered relationship of `Lineage`/`LineageFactory` + current version + Go struct into Thema under some new abstraction (it's sorta like a proto-[kernel](https://github.com/grafana/thema/tree/main/kernel)?) then we can put an assignability check into there, probably even generically in Thema.

Until then, though, it seemed OK to put here. Having it here provides the guarantee, "all  thema schemas in the registry are assignable to their corresponding Go structs," which is ultimately what we want, and also reasonable statement of concern separation to make. It might be better to have earlier in the process, but it's hard to imagine a case in which something escapes this check.

Reply:
Sounds good. I think this is definitely something that we could / should verify either by using a Kernel or with codegen.